Port to Firefox Addon

rpl opened the pull request:
working on fixing #14 

I need to test it a bit more but this is already a working prototype ;-)
(ng-inspector does already a very good job to keep the browser specific addon code to the minimum)

build generated files (stylesheet.css & ng-inspector.js): ```gulp```

## CFX (30 <= Firefox < 38)

Install addon-sdk 1.17 from https://ftp.mozilla.org/pub/mozilla.org/labs/jetpack/jetpack-sdk-latest.zip
(Reference: https://developer.mozilla.org/en-US/Add-ons/SDK/Tutorials/Installation)

Activate cfx (e.g. ```source bin/activate```)

In the ng-inspector.firefox dir:
- How to try it: ```cfx run -b /path/to/firefox-30.0/firefox```
- How to build xpi: ```cfx xpi```

## JPM (Firefox >= 38)

How to try it:

- npm install (it installs mozilla jpm build tool into node_modules/jpm)
- npm run run-xpi -- -b /path/to/firefox-38/firefox

How to build xpi:

- npm run build-xpi

![ng-inspector running on Firefox](https://cloud.githubusercontent.com/assets/11484/7183153/09bc9f24-e456-11e4-93e3-f2d0c210e5e9.png)

rev087 commented:
This is awesome, thanks a lot for the PR @rpl! I have some client work to finish but I'll look into this asap.

rpl commented:
@rev087 thanks to you, the code is so clean that I turned it into a Firefox addon in no time ;-)

just ping me for any question or any needed tweak

DrewML commented on .gitignore:
```diff
@@ -1,3 +1,9 @@
+## generated files
+ng-inspector.firefox/*.xpi
+ng-inspector.firefox/data/icons/
+ng-inspector.firefox/data/ng-inspector.js
```
May want to double-check with @rev087 on ignoring these files. The current pattern for this repository has been to check-in the built files.

rpl replied:
@DrewML no problem from my side to remove these paths from the gitignore and add the generated assets in the pull request once reviewed

I prefer to keep the generated files out of the repo (just to prevent any chance to have the committed versions to be out of sync with their last generated version) but in this case I preferred to do not add them into the branch just to keep it clean to help the review to focus only on the effective differences (and add them later in a separate commit in this branch).

DrewML commented on ng-inspector.firefox/lib/main.js:
```diff
@@ -0,0 +1,42 @@
+  }
+});
+
+var pageModConfig = manifest["firefox-devtools"]["page-mods"][0]
```
Missing a semi-colon here.

DrewML commented on package.json:
```diff
@@ -35,6 +37,7 @@
     "protractor": "~0.24.0",
-    "http-server": "~0.6.1"
+    "http-server": "~0.6.1",
+    "jpm": "^1.0.0"
```
Mozilla mentions [here](https://github.com/mozilla/jpm) that jpm `Currently only works with Firefox 38+`, which is [being released in May](https://developer.mozilla.org/en-US/Firefox/Releases/38). Do you know if there is any way to to automate building of a Firefox extension that would allow `ng-inspector` to support older versions?

rpl replied:
@DrewML I tried to use only stable APIs from the addon-sdk, so from a code "point of view" this should work on older version of Firefox too.

Said that, when jpm builds the xpi archive, it puts a "bootstrap.js" loader inside the xpi which is not compatible with Firefox versione older than 38.

I'm going to ensure this package build correctly even with ```cfx``` (the previous xpi build tool), which generates xpis compatible with older versions.

DrewML commented:
@rpl This is awesome! Thanks for knocking this out.

I took a quick look at the PR and left some initial comments to try and help out @rev087. But this looks awesome!

rpl commented:
@rev087 @DrewML In ac66cc8 I've applied a couple of tweaks to make the Firefox addon compatible (runnable && buildable) with both jpm and cfx and updated the pull request description with instructions about how to build and run it with cfx (nevertheless I can put together a gulp-mozilla-addon-sdk package as I've done with grunt-mozilla-addon-sdk before if we prefer an automated build task with the cfx tool too)

I've created a preview-release build in my forked repo if someone wants to give it a try without build it by himself:
- https://github.com/rpl/ng-inspector/releases/tag/v0.5.8-firefox

rev087 commented:
Great work, thanks @rpl. I'll try to make some time this weekend to go through your PR and publish a Firefox release. I really appreciate that you went the extra mile with the build process as well.

Did you feel the graphics available (toolbar button icon etc.) were enough? If Firefox requires different sizes/styles or additional graphics, I'd be happy to adapt/create them.

rpl commented:
@rev087 Great, thanks! Let me know if you need help or any info.

I've tried to be consistent to the user experience of the Chrome version and it seems to me that the existing graphics assets are appropriate and they looks ok on Firefox.

rpl commented:
@rev087 I've rebased this branch on the new master, because I have fixed a couple of things in the ```ng-inspector.js``` injected script in 9cf9ad3 and then preemptively I have resolved a couple of conflicts with the new upcoming v0.5.10.

- in 65d7858 I've added the caret icon to the src/icons/ (or it will be missing in the Firefox xpi)
- in 193a096 I've fixed an issue with the v.0.5.10 ng-inspector.js building (src/js/utils.js should be src/js/Utils.js)
